### PR TITLE
Issue 49

### PR DESCRIPTION
**Pull request by skad0**
#49 
But with small change due to result order. Block and his mods should be first.

### CODE REVIEW

**Comment by zxqfox** [comment]
For DAAAASSS

File: lib/normalize2.js [hunk 1]
```diff
@@ -102,7 +104,7 @@ module.exports = function (decl) {
     return res;
 
     function add(entity, tech) {
-        const str = naming.stringify(entity) + tech;
+        let str = entity.block + '__' + entity.elem + '_' + entity.modName + '_' + entity.modVal;
```
Comment (zxqfox) [inline lib/normalize2.js]: Ouh. It's strange

Comment (skad0) [inline lib/normalize2.js]: bem-naming ignore cases where block null or undefined which important for us.

File: lib/normalize2.js [hunk 2]
```diff
@@ -1,7 +1,5 @@
 'use strict';
 
-const naming = require('bem-naming');
```
Comment (zxqfox) [inline lib/normalize2.js]: Can we drop it in package.json?

Comment (skad0) [inline lib/normalize2.js]: As i see, it used in other module.

File: lib/normalize2.js [hunk 3]
```diff
@@ -18,19 +16,23 @@ module.exports = function (decl) {
         if (typeof entity === 'string') {
             block = entity;
         } else {
-            block = entity.block;
+            block = entity.block || null;
             elem = entity.elem;
```
Comment (zxqfox) [inline lib/normalize2.js]: Well, I guess this is the same for `elem`.

Could you add some cases without block and elem?

File: lib/normalize2.js [hunk 4]
```diff
@@ -97,12 +104,21 @@ module.exports = function (decl) {
...
+            item.modVal = val;
+            add(item, tech);
+        }
     }
 
     return res;
 
     function add(entity, tech) {
-        const str = naming.stringify(entity) + tech;
+        const str = entity.block + '__' + entity.elem + '_' + entity.modName + '_' + entity.modVal;
```
Comment (tadatuta) [inline lib/normalize2.js]: Let's use something like this
```js
const key = Object.keys(entity).join();
```
instead to make it super clear we don't need proper BEM naming here but just any uniq key for each entity.

File: lib/normalize2.js [hunk 5]
```diff
@@ -97,12 +104,21 @@ module.exports = function (decl) {
                 }
             }
         }
+
+        if (isNotActual(mod) && val) {
+            let item = {};
```
Comment (tadatuta) [inline lib/normalize2.js]: shouldn't it be `const`?

File: lib/normalize2.js (hunk 4 above)
Comment (zxqfox) [inline lib/normalize2.js]: Object.values?

Comment (skad0) [inline lib/normalize2.js]: Then, maybe something like this?
`Object.keys(entity).reduce((prev, key) => { return prev + entity[key] }, '') + tech;`

Comment (tadatuta) [inline lib/normalize2.js]: 👍

File: test/assign.test.js [hunk 6]
```diff
@@ -167,3 +167,23 @@ test('should use modVal with scope if only tech given', t => {
         { entity: { block: 'sb', modName: 'sm', modVal: 'sv' } }),
         { entity: { block: 'sb', modName: 'sm', modVal: 'sv' }, tech: 'bemhtml' });
 });
+
+test('should use scope vals if null given', t => {
+    t.deepEqual(
+        assign(
+            { entity: { block: null, modName: 'mod', modVal: 'val' } },
+            { entity: { block: 'block', elem: 'elem' }, tech: 'bemhtml' }
```
Comment (zxqfox) [inline test/assign.test.js]: здесь явно elem и tech лишние

File: test/normalize2/elems-mods.test.js [hunk 7]
```diff
@@ -65,3 +66,18 @@ test('should support mods in elems and block', t => {
...
+test('should support block mods with `elems` field without block', t => {
+    const decl = [
+        {
+            elems: ['close'],
+            mods: { theme: 'protect' }
+        }
+    ];
+
+    t.deepEqual(normalize(decl), [
+        { entity: { block: null }, tech: undefined },
```
Comment (Yeti-or) [inline test/normalize2/elems-mods.test.js]: guys why it normalize to `{ entity: { block: null }, tech: undefined },` is it really usefull ?
who needs empty block with empty tech ?

Comment (skad0) [inline test/normalize2/elems-mods.test.js]: I think, idea is that we need some block, but we don't know which on this step. After normalize we call assign and get this information from scope.

**Comment by zxqfox** [comment]
LGTM :shipit:

**Comment by Yeti-or** [comment]
🎱

**Pull request merged**